Hsin Yi iP

hsinyilow opened the pull request:
### Astra :sparkles:

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Astra helps you keep track of the many ideas you may have.
It is
- _fast_
- lightweight
- **free**

To use it, you just need to:
1. Download it from [here](https://github.com/hsinyilow/ip/)
2. Double click it
3. Add your tasks

Features:
- [x] Marking tasks
- [x] Shows date for deadline and events
- [ ] Manages deadlines

You can use it to practice Java too! Here is a snippet of the `command` method
``` ruby
    public void command(String input) {
        if (input.equals("list")) {
            //show list
            displayTaskList();
        } else if (input.startsWith("delete")) {
            deleteTask(input);
        }
        ...
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

hsinyilow committed:
Removed default Duke, renamed Duke to Astra. Added in greeting and exit.

hsinyilow committed:
Other files to commit for level-0 after renaming.

hsinyilow committed:
Chatbot echos command entered by user, exits using command "bye"

hsinyilow committed:
store text and display them as a list when requested

hsinyilow committed:
Created a new class for task. Add the ability to to mark/unmark tasks as done.

hsinyilow committed:
Added todo, deadline and event tasks

hsinyilow committed:
Use the input/output redirection technique to semi-automate the testing of Duke. Fixed small output bug on adding new task.

hsinyilow committed:
Handled invalid input. Restructured Tasks and its subclass to better handle invalid input.

hsinyilow committed:
Added option to delete tasks from the list.

hsinyilow committed:
Add skeleton of new save system

hsinyilow committed:
Add save and load functionalities to save system

hsinyilow committed:
Merge branch 'branch-Level-7'

hsinyilow committed:
Update tasks to recognise date and time command

Deadline and Event tasks recognise everything as a string.

Date and time recognising is more useful for future functionalities.

Let's,
* update the parser to recognise date and time
* save date and time as LocalDateTime

hsinyilow committed:
Update save system to support date time saving

hsinyilow committed:
Merge branch 'branch-Level-8'

hsinyilow committed:
Refactor code to follow OOP rules

Not all classes follow OOP rules.

Better for classes to follow OOP for code reusability and flexibility.

Let's:
*Change Task.java to abstract class
*Add Ui, Parser, TimeData classes
*Move functions to appropriate classes

hsinyilow committed:
Merge branch 'branch-Increments'

hsinyilow committed:
Move classes into suitable java packages

hsinyilow committed:
Merge branch 'branch-Increments'

hsinyilow committed:
Merge branch 'add-gradle-support'

hsinyilow committed:
Add checkstyle to gradle

hsinyilow committed:
Merge branch 'branch-Increments'

hsinyilow committed:
Add some JUnit tests

hsinyilow committed:
Merge branch 'branch-Increments'

hsinyilow committed:
Update build settings to build jar file

hsinyilow committed:
Merge branch 'branch-Increments'

hsinyilow committed:
Add find functionality to tasklist

hsinyilow committed:
Update all files to match coding standard

hsinyilow committed:
Add JavaDoc for most functions

hsinyilow committed:
Revert "Add JavaDoc for most functions"

This reverts commit 8fe2b0547db156076c4ad55fa8d967cb33e21db0.

hsinyilow committed:
Merge branch 'branch-A-CodingStandard'

hsinyilow committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/astra/task/Task.java
#	src/main/java/astra/task/TaskList.java

hsinyilow committed:
Update JavaDoc for all classes

hsinyilow committed:
Merge branch 'branch-A-JavaDocNew'

# Conflicts:
#	src/main/java/astra/system/AstraException.java
#	src/main/java/astra/system/Parser.java
#	src/main/java/astra/system/SaveSystem.java
#	src/main/java/astra/system/TimeData.java
#	src/main/java/astra/system/Ui.java
#	src/main/java/astra/task/Task.java
#	src/main/java/astra/task/TaskList.java

hsinyilow committed:
Merge branch 'branch-A-JavaDocNew'

hsinyilow committed:
Fix coding standard violations after branch merge

hsinyilow committed:
Add check style and fix coding standard violations

hsinyilow committed:
Merge branch 'A-CheckStyle'

hsinyilow committed:
Import javaFX

kubrian commented on src/main/java/astra/Astra.java.rej:
I think this file [should be removed](https://stackoverflow.com/questions/34585865/what-are-rej-files-which-are-created-during-merge)

kubrian commented on src/test/java/astra/system/ParserTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Test
+    public void testCommand1() {
+        assertEquals("aee", Parser.parseCommand("       a     e     e   ", 2, true));
+    }
```
I think a more descriptive name, as well as a comment to explain the string will make the purpose of test case more intuitive.

kubrian commented on src/main/java/astra/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+package astra.task;
+
+/**
+ * Is inherited by all tasks.
```
Consider using "Represents an abstract task that ..."

kubrian reviewed (commented):
LGTM

hsinyilow committed:
Add GUI by following JavaFX tutorial part 2

hsinyilow committed:
Update GUI for Astra according to JavaFX tutorial

hsinyilow committed:
Update files to match coding standard

hsinyilow committed:
Merge branch 'branch-Level-10'

DanielJames0302 commented on src/main/java/astra/Astra.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Command and result loop
+     */
+    public void runAstra() {
```
Good method naming

DanielJames0302 commented on src/main/java/astra/system/Parser.java:
```diff
@@ -0,0 +1,85 @@
+            LocalTime time = LocalTime.MIN;
+            if (parseInput.length != 1) {
+                time = LocalTime.of(Integer.parseInt(parseInput[1].substring(0, 2)),
+                        Integer.parseInt(parseInput[1].substring(2)));
```
Well aware of the indentation

DanielJames0302 reviewed (commented):
LGTM. Well done.

hsinyilow committed:
Update and refactor code to improve code quality

Code does not follow code quality standards.
It also has coding standard violations.

Updating and refactoring the code allows the code to be more readable,
and follows the standard that was provided.

Let's,
- update the boolean names
- have a static method to handle initializing of tasks objects for
  event, deadline and todo tasks
- shift some parts of addTask function to command function

This is done as:
- better boolean names will fit the coding standard (use of "is")
- task constructors does not have an obvious happy path and it makes
  better sense for the parsing to be done in another function, and for
  the constructor to only handle assignments of variables
- the else block in the command should be used to catch error instead
  of leaving command errors to be handled by the add function [code
  quality violation]

hsinyilow committed:
Add assertions to Parser and SaveSystem classes

hsinyilow committed:
Add assertions to Task child classes

There are no assertions to define assumptions about the input
parameters for the Task child public constructor

Any classes can call the constructors as they are public, but the
constructors had assumed that they are called for the right purpose.
Wrong calls would fail for the wrong errors.

Add assertions to the constructors to check if the assumptions hold.

It is better to have assertions because it will ensure the functions
input parameters follows the assumption.

hsinyilow committed:
Merge pull request #1 from hsinyilow/branch-A-Assertions

Add assertions to the project from branch 'A-Assertions'

hsinyilow committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/astra/system/Parser.java
#	src/main/java/astra/system/SaveSystem.java
#	src/main/java/astra/task/DeadlineTask.java
#	src/main/java/astra/task/EventTask.java
#	src/main/java/astra/task/TodoTask.java

hsinyilow committed:
Merge pull request #2 from hsinyilow/branch-A-CodeQuality

Improve code quality of Astra

hsinyilow committed:
Add update task functionality

No way to update a task detail unless user deletes and adds it again.

This causes inconvenience to the user. Also, since we can update
mark within the system, detail should also be allowed to update within
the system.

Let's,
- add an update command that allows user to select a specific task to
  modify the details.
- add an updateTask function to TaskList to get the specific task
- add an updateDetails function to each of the Task child classes to
  handle update for each of their own type of details field.

Splitting the implementation this way is to maintain the abstraction
and encapsulation properties of OOP programming for the classes.

hsinyilow committed:
Merge branch 'branch-Edits'

hsinyilow committed:
Update gui and fix time parsing bug

hsinyilow committed:
Update JavaDoc comments

hsinyilow committed:
Add user guide and product screenshot

hsinyilow committed:
Add Junit tests and fix saving bug

hsinyilow committed:
Update files to match coding standards

hsinyilow committed:
Update user guide and junit tests

hsinyilow committed:
Update file to match coding standards
